engine/src/main/java/org/pentaho/di/trans/Trans.java

```diff
@@ -0,0 +1,5804 @@
+            .getLinesUpdated(), step.getLinesRejected(), step.getErrors() );
+
+        synchronized ( stepPerformanceSnapShots ) {
+          List<StepPerformanceSnapShot> snapShotList = stepPerformanceSnapShots.get( step.toString() );
+          StepPerformanceSnapShot previous;
+          if ( snapShotList == null ) {
+            snapShotList = new ArrayList<>();
+            stepPerformanceSnapShots.put( step.toString(), snapShotList );
+            previous = null;
+          } else {
+            previous = snapShotList.get( snapShotList.size() - 1 ); // the last one...
+          }
+          // Make the difference...
+          //
+          snapShot.diff( previous, step.rowsetInputSize(), step.rowsetOutputSize() );
+          snapShotList.add( snapShot );
+
+          if ( stepPerformanceSnapshotSizeLimit > 0 && snapShotList.size() > stepPerformanceSnapshotSizeLimit ) {
+            snapShotList.remove( 0 );
+          }
+        }
+      }
+
+      lastStepPerformanceSnapshotSeqNrAdded = stepPerformanceSnapshotSeqNr.get();
+    }
+  }
+
+  /**
+   * This method performs any cleanup operations, typically called after the transformation has finished. Specifically,
+   * after ALL the slave transformations in a clustered run have finished.
+   */
+  public void cleanup() {
+    // Close all open server sockets.
+    // We can only close these after all processing has been confirmed to be finished.
+    //
+    if ( steps == null ) {
+      return;
+    }
+
+    for ( StepMetaDataCombi combi : steps ) {
+      // PDI-18214/CDA-243: Check if the steps have been disposed
+      if ( !combi.data.isDisposed() ) {
+        combi.step.setOutputDone();
+        combi.step.dispose( combi.meta, combi.data );
+        combi.step.markStop();
+      }
+      combi.step.cleanup();
+    }
+  }
+
+  /**
+   * Logs a summary message for the specified step.
+   *
+   * @param si the step interface
+   */
+  public void logSummary( StepInterface si ) {
+    log.logBasic( si.getStepname(), BaseMessages.getString( PKG, "Trans.Log.FinishedProcessing", String.valueOf( si
+      .getLinesInput() ), String.valueOf( si.getLinesOutput() ), String.valueOf( si.getLinesRead() ) ) + BaseMessages
+      .getString( PKG, "Trans.Log.FinishedProcessing2", String.valueOf( si.getLinesWritten() ), String.valueOf( si
+        .getLinesUpdated() ), String.valueOf( si.getErrors() ) ) );
+  }
+
+  /**
+   * Waits until all RunThreads have finished.
+   */
+  public void waitUntilFinished() {
+    try {
+      if ( transFinishedBlockingQueue == null ) {
+        return;
+      }
+      boolean wait = true;
+      while ( wait ) {
+        wait = transFinishedBlockingQueue.poll( 1, TimeUnit.DAYS ) == null;
+        if ( wait ) {
+          // poll returns immediately - this was hammering the CPU with poll checks. Added
+          // a sleep to let the CPU breathe
+          Thread.sleep( 1 );
+        }
+      }
+    } catch ( InterruptedException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)